lib/worker/OperationQueue.js

```diff
@@ -0,0 +1,363 @@
+/*!
+ * Copyright (c) 2017-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const _continuityConstants = require('../continuityConstants');
+const _util = require('../util');
+const database = require('bedrock-mongodb');
+const logger = require('../logger');
+const pLimit = require('p-limit');
+
+// Controls the max concurrency for a given number of promises executing. This
+// is done to ensure we reduce pressure on the NodeJS Promise Queue and limit
+// the number concurrent I/O operations. The number 25 was arbitrarily picked.
+const limit = pLimit(25);
+// Controls the maximum number of documents to pull from the database and cache
+// locally. The Math.max(DEFAULT_MAX_CACHED_QUEUE_SIZE, MAX_OPS_PER_REG_EVENT)
+// will be the final value used to limit the number of documents pulled from
+// the database.
+const DEFAULT_MAX_CACHED_QUEUE_SIZE = 250;
+
+// Error Messages
+const OPERATIONS_REMOVAL_ERROR = 'Failed to remove operations from queue.';
+const OPERATION_VALIDATION_ERROR = 'Non-critical operation validation error. ' +
+  'Invalid head of operation queue.';
+
+module.exports = class OperationQueue {
+  constructor({worker}) {
+    // continuity worker
+    this.worker = worker;
+    const {ledgerNode} = worker;
+    // ledger node instance
+    this.ledgerNode = ledgerNode;
+    // ledger node's id
+    this.ledgerNodeId = ledgerNode.id;
+    // operation queue collection
+    this.collection = database.collections.continuity2017_operation_queue;
+    // cached chunk of operations to return for a regular event
+    this.chunk = null;
+    // tracks whether or not there are more operations after the next chunk
+    this.hasMore = false;
+    // cached queue from the database
+    this.queue = null;
+    // index of the head of the cached queue of operations
+    this.head = 0;
+    // count of possible operations to be added in the chunk
+    this.opCount = 0;
+    // list of operations to be removed
+    this.removedOperations = [];
+    // caches operation removal promise
+    this.removing = null;
+    // caches database empty check promise
+    this.databaseEmpty = null;
+    // caches fetch record promise
+    this.fetchRecords = null;
+    // constants
+    this.MAX_OPERATIONS = _continuityConstants.events.maxOperations;
+    this.MAX_CACHED_QUEUE_SIZE = Math.max(
+      DEFAULT_MAX_CACHED_QUEUE_SIZE,
+      this.MAX_OPERATIONS);
+  }
+
+  /**
+   * Returns whether or not there is another chunk of operations that can
+   * be put into a local regular event. This MUST NOT be called concurrently.
```

Comment:
`or not` is superfluous.

```suggestion
   * Returns whether there is another chunk of operations that can be put
   * into a local regular event. This MUST NOT be called concurrently.
```